[FLINK-12652] [documentation] add first version of a glossary (#8607)
apache/flink: docs/concepts/glossary.md

---
title: Glossary
nav-pos: 3
nav-title: Glossary
nav-parent_id: concepts
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

#### Flink Application Cluster

A Flink Application Cluster is a dedicated [Flink Cluster](./glossary#flink-cluster) that only
executes a single [Flink Job](./glossary#flink-job). The lifetime of the
[Flink Cluster](./glossary#flink-cluster) is bound to the lifetime of the Flink Job. Formerly

> Flink Application Clusters were also known as Flink Clusters in *job mode*.

Author: Done

#### Flink Cluster

The distributed system consisting of (typically) one Flink Master process and one or more Flink

> Taskmanagers processes.

Author: In my opinion, we should use "Flink Master (process)" instead of Flink JobManager, yes. Flink JobManager is too overloaded, since the FLIP-6 effort. Are there cases, where the JobManagers are not running in the same process as Resource Manager and Dispatcher?

I also thought about adding these to the Glossary. We should also add JobManager then, to avoid confusion, so: Flink Master, JobManager, TaskManager. 

What do you think?

> Taskmanagers processes.

Reviewer: In theory, yes: JobManager, ResourceManager and Dispatcher can be deployed separately...Flink's own startup scripts are not (yet?) making use of this though, afaik.

Unless you want to introduce new terminology, I think, we should stick with JobManager (process) even though it may contain more components.

> Taskmanagers processes.

Author: I think, I would like to introduce new terminology ;) If I understood @StephanEwen correctly, he was also in favor of new terminology, but it is of course tricky.

> Taskmanagers processes.

Author: @NicoK I changed this for "JobManager Process" for now. Let's move the discussion about renaming this to the appropriate PR for the Concepts or Deployment Section.

> Taskmanagers processes.

Reviewer 2: FWIW, I like `Flink Master Process` for the reasons outlined by Konstantin. I think it's a good call to move the discussion of whether we want to change it to another place/time.

> Taskmanagers processes.

Reviewer 3: just FYI, I also used "Flink master process" when describing the concept in the book.

#### Event

> An event is a statement of fact about a change of the state of the domain modelled by the

Reviewer: "Statement or fact"?

Author: You're right this sounds too complicated. I changed it to just "statement".

Reviewer: ok, I first assumed a typo: `of`->`or`...but but it was really not. Then you're right :)

application. Events can be input and/or output of a stream or batch processing application.
Events are special types of [records](./glossary#Record)

#### ExecutionGraph

see [Physical Graph](./glossary#physical-graph)

#### Function

Functions, or user-defined functions (UDFs), are implemented by the user and encapsulate the
application logic of a Flink program. Most Functions are wrapped by a corresponding
[Operator](./glossary#operator).

#### Instance

The term *instance* is used to describe a specific instance of a specific type (usually
[Operator](./glossary#operator) or [Function](./glossary#function)) during runtime. As Apache Flink
is mostly written in Java, this corresponds to the definition of *Instance* or *Object* in Java.

> In the context of Apache Flink, the term *parallel instance* is also frequently used to emphasizes

Reviewer: "to emphasize"

Author: Done

that multiple instances of the same [Operator](./glossary#operator) or
[Function](./glossary#function) type are running in parallel.

#### Flink Job

> A Flink Job is the runtime representation of a Flink program. A Flink Job can be either submitted

Reviewer: "can either be..."

Author: Done.

to a long running [Flink Session Cluster](./glossary#flink-session-cluster) or it can be started as a
self-contained [Flink Application Cluster](./glossary#flink-application-cluster).

#### JobGraph

see [Logical Graph](./glossary#logical-graph)

#### Logical Graph

A logical graph is a directed graph describing the high-level logic of a stream processing program.
The nodes are [Operators](./glossary#operator) and the edges indicate input/output-relationships or
data streams or data sets.

#### Managed State

> Managed State describes application state, which has been registered with the framework. For

Reviewer: No comma before `which`

Author: Done

> Managed State Apache Flink will take care about persistence and rescaling among other things.

Reviewer: "For Managed State,..." (here we need the comma)

Author: Done

#### Operator

Node of a [Logical Graph](./glossary#logical-graph). An Operator performs a certain operation,
which is usually executed by a [Function](./glossary#function). Sources and Sinks are special
Operators for data ingestion and data egress.

#### Partition

A partition is an independent subset of the overall data stream or data set. A data stream or
data set is divided into partitions by assigning each [record](./glossary#Record) to one or more
partitions. Partitions of data streams or data sets are consumed by [Tasks](./glossary#task) during
runtime.

#### Physical Graph

A physical graph is the result of translating a [Logical Graph](./glossary#logical-graph) for
execution in a distributed runtime. The nodes are [Tasks](./glossary#task) and the edges indicate

> input/output-relationships or partitions of data streams or data sets.

Reviewer: link partition?

Author: Done

#### Record

Records are the constituent elements of a data set or data stream.
[Operators](./glossary#operator) and [Functions](./glossary#Function) receive records as input
and emit records as output.

#### Flink Session Cluster

> A long-running [Flink Cluster](./glossary#flink-cluster), which accepts multiple

Reviewer: No comma

Author: Done

[Flink Jobs](./glossary#flink-job) for execution. The lifetime of this Flink Cluster is not bound
to the lifetime of any Flink Job. Formerly, a Flink Session Cluster was also known as a Flink Cluster in
*session mode*.

#### State Backend

> For stream processing programs the State Backend determines how state is stored on each Taskmanager

Reviewer: "For stream processing programs,..." (with comma)

Author: Done

> (Java Heap of Taskmanager or RocksDB) as well as where it is written upon a checkpoint (Java Heap of Flink Master or

Reviewer: Maybe also emphasize "(embedded) RocksDB on TaskManager"?

Author: Done

Filesystem).

#### Sub-Task

> A Sub-Task is a [Task](./glossary#task) responsible for processing a

Reviewer: "is an instance of a Task"?

Author: Done

[partition](./glossary#partition) of the data stream. The term "Sub-Task" emphasizes that there are

> multiple parallel Tasks for the same [Operator](./glossary#operator).

Reviewer: "operator(s)"

Author: Done

#### Task

Node of a [Physical Graph](./glossary#physical-graph). A task is the basic unit of work, which is
executed by Flink's runtime. Tasks usually encapsulate one or more [Operators](./glossary#operator).

#### Transformation

A transformation is applied on one or more data streams or data sets and results in one or more
output new data streams or data sets. A transformation might change a data stream or data set on a
per-record basis, but but might also only change its partitioning or perform an aggregation. In
contrast to operations, transformations operate on streams or data sets, while operations operate

> on [records](./glossary#Record).

Reviewer: I don't really get the last argument. Also, do you compare it with Operators, or Operations?

Author: For me, an operation would be the action that an operator performs. 

The idea is that transformations are applied on streams, while operations are applied on individual records.

Reviewer: Can you give an example of a "transformation" and an "operation"? Maybe I'm mixing things up here...

Author: A transformation and and operation can do exactly the same thing. For example, a simple Map can be an operation if you talk about it on a per-record basis and a transformation if you talk about it on a stream.

For a repartitioning operations it makes a bit more sense, I think. A `keyBy` results is a transformation of the stream, but you can not repartition a single record. Of course, you could view it as a per-record operation, but it makes less sense, I think.

Reviewer: That makes things clearer, but do we actually need the two different terms? I think, it may actually confuse people (like me) - unless more specifically explained as here.

Author: I changed this now. Since we don't mention the term "operation" anywhere, I removed this sentence and just noted that a transformation usually, but not always corresponds to an operator.

to a long running [Flink Session Cluster](./glossary#flink-session-cluster) or it can be started as a
self-contained [Flink Application Cluster](./glossary#flink-application-cluster).

#### JobGraph

see [Logical Graph](./glossary#logical-graph)

#### JobManager

JobManagers are one of the components which make up the [Flink Master](./glossary#flink-master). A
JobManager is responsible for supervising the execution of the [Tasks](./glossary#task) of a single
job.

#### Logical Graph

A logical graph is a directed graph describing the high-level logic of a stream processing program.
The nodes are [Operators](./glossary#operator) and the edges indicate input/output-relationships or
data streams or data sets.

#### Managed State

Managed State describes application state which has been registered with the framework. For
Managed State, Apache Flink will take care about persistence and rescaling among other things.

#### Flink Master

The Flink Master is the master process of a [Flink Cluster](./glossary#flink-cluster). It consists
of three main components: resource manager, dispatcher and one [JobManager](./glossary#job-manager)

> per running [Flink Job](./glossary#flink-job).

Reviewer: irrespective of whether to use "JobManager" or "Flink Master", reading about 3 components and then having "on JM per job" sounds a bit strange

Author: three "different" components insteads?

Reviewer: yes, that might work

Author: Done.

---
title: Glossary
nav-pos: 3
nav-title: Glossary
nav-parent_id: concepts
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

> #### Flink Application Cluster

Reviewer 2: Personally, I call these `job cluster` in line with the internals, e.g. [JobClusterEntrypoint](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/JobClusterEntrypoint.java).

cc @tillrohrmann

Reviewer 4: It is the same for me. I'm also fine with changing the terminology as long as we don't confuse our users.

Author: The term "Flink Application (Cluster)" was proposed by @StephanEwen to emphasize, that you basically run a Flink job like an classical application in this mode (or at least this is the mid-term goal (passive resource manager, Flink as a library). 

@tillrohrmann @uce Do you think we can change this at this point?

Reviewer 2: I do like the proposed name and I'd be happy to call it application cluster going forward myself.

I think having slight misalignment with internal class naming is acceptable, but I'd suggest to update user facing references to job cluster if we keep the renaming:
* https://ci.apache.org/projects/flink/flink-docs-release-1.8/ops/deployment/kubernetes.html#flink-job-cluster-on-kubernetes
* https://github.com/apache/flink/tree/master/flink-container/docker
* https://github.com/apache/flink/tree/master/flink-container/kubernetes

Otherwise I would be afraid of ending up in a situation with inconsistent names which can be confusing to users.

Since I am not involved in the day to day development of this component, I would leave the decision to @tillrohrmann whether we can change it at this point.

Author: Agreed. There are actually a few changes "planned" to this module anyway (see discussion of https://issues.apache.org/jira/browse/FLINK-12546), so this might actually be good timing to rename it.

Author: @tillrohrmann is fine to change this. @uce I created https://issues.apache.org/jira/browse/FLINK-12885 to rename Job Cluster to Application Cluster. Shouldn't be a problem to do this before the 1.9 release.

Author: I have already a local version of FLINK-12885 now, but it is still blocked on a different issue (FLINK-12752). Once this is merged, we can align this pretty quickly.

A Flink Application Cluster is a dedicated [Flink Cluster](./glossary#flink-cluster) that only
executes a single [Flink Job](./glossary#flink-job). The lifetime of the
[Flink Cluster](./glossary#flink-cluster) is bound to the lifetime of the Flink Job. Formerly
Flink Application Clusters were also known as Flink Clusters in *job mode*. Compare to
[Flink Session Cluster](./glossary#flink-session-cluster).

#### Flink Cluster

The distributed system consisting of (typically) one Flink Master process and one or more Flink
Taskmanagers processes.

#### Event

An event is a statement about a change of the state of the domain modelled by the
application. Events can be input and/or output of a stream or batch processing application.

> Events are special types of [records](./glossary#Record)

Reviewer 3: Add `.` at the end

#### ExecutionGraph

see [Physical Graph](./glossary#physical-graph)

#### Function

> Functions, or user-defined functions (UDFs), are implemented by the user and encapsulate the

Author: Good point. Done.

#### Flink Job

A Flink Job is the runtime representation of a Flink program. A Flink Job can either be submitted
to a long running [Flink Session Cluster](./glossary#flink-session-cluster) or it can be started as a
self-contained [Flink Application Cluster](./glossary#flink-application-cluster).

#### JobGraph

see [Logical Graph](./glossary#logical-graph)

#### Flink JobManager

JobManagers are one of the components running in the
[Flink JobManger Process](./glossary#flink-jobmanager-process). A JobManager is responsible for
supervising the execution of the [Tasks](./glossary#task) of a single job.

#### Logical Graph

A logical graph is a directed graph describing the high-level logic of a stream processing program.
The nodes are [Operators](./glossary#operator) and the edges indicate input/output-relationships or
data streams or data sets.

#### Managed State

Managed State describes application state which has been registered with the framework. For
Managed State, Apache Flink will take care about persistence and rescaling among other things.

#### Flink JobManager Process

The Job Manager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called

> *JobManager* for historical reasons, but actually has actually contains three distinct components:

Author: Currently, they are always run together, correct? I assume the other option would be: 

* Flink Master Process: Dispatcher &  Resource Manager
* One JobManager Process per JobManager separately

How about we it like it is right now and revisit this, when we have decided about introducing the term "Flink Master Process" in the course of the FLIP?

> *JobManager* for historical reasons, but actually has actually contains three distinct components:

Reviewer 2: SGTM.

> *JobManager* for historical reasons, but actually has actually contains three distinct components:

Reviewer 3: AFAIK, they are not always run together. At least that's what is described by FLIP-6. But I don't want to nitpick on this

> *JobManager* for historical reasons, but actually has actually contains three distinct components:

Author: @fhueske In which case are these not run together? AFAIK also in YARN/Mesos session mode, there is only one Master Process for all Job Managers, although conceptually they would not have to run together (since FLIP-6).

> *JobManager* for historical reasons, but actually has actually contains three distinct components:

Author: I talked to @tillrohrmann offline and they are currently always run together in the same process + there are no plans to change this in the foreseeable future.

Flink Resource Manager, Flink Dispatcher and one [Flink JobManager](./glossary#flink-jobmanager)
per running [Flink Job](./glossary#flink-job).

#### Operator

Node of a [Logical Graph](./glossary#logical-graph). An Operator performs a certain operation,
which is usually executed by a [Function](./glossary#function). Sources and Sinks are special
Operators for data ingestion and data egress.

#### Operator Chain

> An Operator Chain consists of one or more consecutive [Operators](./glossary#operator) without any

Reviewer 3: I would not call a single operator an operator chain. IMO a chain consist of at least two operators.

Author: Ok, personally I think a chain of size 1 is perfectly fine, but I'll change it ;)

called repartitioning.

#### Physical Graph

A physical graph is the result of translating a [Logical Graph](./glossary#logical-graph) for
execution in a distributed runtime. The nodes are [Tasks](./glossary#task) and the edges indicate
input/output-relationships or [partitions](./glossary#partition) of data streams or data sets.

#### Record

Records are the constituent elements of a data set or data stream.
[Operators](./glossary#operator) and [Functions](./glossary#Function) receive records as input
and emit records as output.

#### Flink Session Cluster

A long-running [Flink Cluster](./glossary#flink-cluster) which accepts multiple
[Flink Jobs](./glossary#flink-job) for execution. The lifetime of this Flink Cluster is not bound
to the lifetime of any Flink Job. Formerly, a Flink Session Cluster was also known as a Flink Cluster in
*session mode*. Compare to [Flink Application Cluster](./glossary#flink-application-cluster).

#### State Backend

For stream processing programs, the State Backend determines how state is stored on each Taskmanager
(Java Heap of Taskmanager or (embedded) RocksDB) as well as where it is written upon a checkpoint

> (Java Heap of Flink Master or Filesystem).

Reviewer 2: ```suggestion
(Java Heap of Flink JobManager or Filesystem).
```

Author: Done

#### Sub-Task

A Sub-Task is an instance of [Task](./glossary#task) responsible for processing a
[partition](./glossary#partition) of the data stream. The term "Sub-Task" emphasizes that there are
multiple parallel Tasks for the same [Operator(s)](./glossary#operator).

#### Task

Node of a [Physical Graph](./glossary#physical-graph). A task is the basic unit of work, which is
executed by Flink's runtime. Tasks encapsulate exactly one parallel

> [Operator Chain](./glossary#operator-chain).

Reviewer 3: "one parallel Operator Chain" -> "one parallel instance of an Operator or Operator Chain"?

Author: Done.

#### Flink TaskManager

TaskManagers are the worker processes of a [Flink Cluster](./glossary#flink-cluster).

> [Tasks](./glossary#task) are scheduled to TaskManagers for execution.

Reviewer 3: Mention that TMs communicate with each other to exchange data between tasks?

Author: Done.

> #### Transformation

Reviewer 3: Mention that a Transformation is an API concept?

Author: Can you elaborate?

Reviewer 3: From the description I read that the term "Transformation" is used in the context of the API, i.e., when implementing an application. When the application is executed it is rather operator or task. If this distinction is not intended, how would a transformation relate to operator/task/function?

Author: Okay, I see. I have added a sentence. Let me know, if this fits.

or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

#### Flink Application Cluster

A Flink Application Cluster is a dedicated [Flink Cluster](#flink-cluster) that only
executes a single [Flink Job](#flink-job). The lifetime of the
[Flink Cluster](#flink-cluster) is bound to the lifetime of the Flink Job. Formerly
Flink Application Clusters were also known as Flink Clusters in *job mode*. Compare to
[Flink Session Cluster](#flink-session-cluster).

#### Flink Cluster

A distributed system consisting of (typically) one

> [Flink JobManager Process](#flink-jobmanager-process) and one or more

Reviewer 5: I have often called this one "master" process, consisting typically of ResourceManager and one or more JobManagers (one per job) and optionally a dispatcher (for session clusters).

For an application cluster: Maser = RM + JM
For as session cluster: Master = RM + Dispatcher + [JM per job]

Reviewer 5: I am not sure is "master" process is the best name. It is similar to the old JobManager, but the name is now a bit misleading.

Author: We discussed calling it "Flink Master Process", but this would introduce a lot of inconsistencies through out the documentation. So, we decided to call it Flink JobManager Process for now and revisit this/change it, when we rework the Deployment section of the documentation.

Reviewer 5: I an skeptical whether terms coined in the glossary ever change easily.

We could also do the following:
  - The per-job component inside the master is actually the "JobMaster" in the code.
  - We could call the process as a while (with RM / Disp. / JobMasters) the "Flink Master Process" and redirect "JobManager" there, mentioning that it was the old term used synonymously.
  - Or we keep the term "JobManager" indefinitely and hope that between JobManager and JobMaster there is not too much confusion.

Author: I think, we have three options: 

1. Rename "JobManager Process" to "Flink Master" now in the glossary. This would be in line with the future vision, but will create a lot of inconsistencies (e.g. naming of scripts and commands in the binary distribution), which we can not fix as part of the glossary. Redirect can mitigate this a bit.

2. Stick with "JobManager Process" right now and rename later consistently. 

3. Stick with "JobManager" forever and rename the component to "JobMaster" in accordance to the code. I think, it is too late for this (all FLIP-6 documents talk about "JobManager") and the names would also not be very descriptive and very similar.

I am happy with 1. and 2., but not so much with option 3. 

@tillrohrmann, @StephanEwen: What do you think?

Reviewer 4: I would be in favor of 1. with redirecting deprecated terms to it + creating follow up issues to fix the terminology in the other docs + code.

Reviewer 5: My preference would be option (1) as well.

Author: Perfect. I'll update the PR accordingly and create a follow-up ticket.

Author: Pushed the change and created https://issues.apache.org/jira/browse/FLINK-12639 .

[Function](./glossary#function) type are running in parallel.

#### Flink Job

A Flink Job is the runtime representation of a Flink program. A Flink Job can either be submitted
to a long running [Flink Session Cluster](./glossary#flink-session-cluster) or it can be started as a
self-contained [Flink Application Cluster](./glossary#flink-application-cluster).

#### JobGraph

see [Logical Graph](./glossary#logical-graph)

#### Flink JobManager

JobManagers are one of the components running in the
[Flink JobManger Process](./glossary#flink-jobmanager-process). A JobManager is responsible for
supervising the execution of the [Tasks](./glossary#task) of a single job.

#### Logical Graph

A logical graph is a directed graph describing the high-level logic of a stream processing program.
The nodes are [Operators](./glossary#operator) and the edges indicate input/output-relationships or
data streams or data sets.

#### Managed State

Managed State describes application state which has been registered with the framework. For
Managed State, Apache Flink will take care about persistence and rescaling among other things.

#### Flink JobManager Process

> The Job Manager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called

Reviewer 2: ```suggestion
The JobManager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called
```

Author: Done.

> *JobManager* for historical reasons, but actually has actually contains three distinct components:

Reviewer 2: ```suggestion
*JobManager* for historical reasons, but actually contains three distinct components:
```

Author: Done.

distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

#### Flink Application Cluster

A Flink Application Cluster is a dedicated [Flink Cluster](./glossary#flink-cluster) that only
executes a single [Flink Job](./glossary#flink-job). The lifetime of the
[Flink Cluster](./glossary#flink-cluster) is bound to the lifetime of the Flink Job. Formerly
Flink Application Clusters were also known as Flink Clusters in *job mode*. Compare to
[Flink Session Cluster](./glossary#flink-session-cluster).

#### Flink Cluster

The distributed system consisting of (typically) one Flink Master process and one or more Flink

> Taskmanagers processes.

Reviewer: `Taskmanagers processes` -> `TaskManager processes`?

actually, the capitalisation should be unified throughout the glossary, sometimes you use `Taskmanager` instead)

Author: Done.

#### Event

An event is a statement about a change of the state of the domain modelled by the
application. Events can be input and/or output of a stream or batch processing application.

> Events are special types of [records](./glossary#Record).

Reviewer: I find this rather complex: `a change of the state of the domain modelled by the
application` and honestly don't know what that means. Maybe rephrase but also give an example (not just "special type or record")

Author: In Domain Driven Design, Event-Driven Architectures, etc. this is exactly, what an event is. It is always an event (=state change) in the domain. I could add a text book example like "checkout" or "item added to cart". Would this help?

#### Function

Functions are implemented by the user and encapsulate the
application logic of a Flink program. Most Functions are wrapped by a corresponding
[Operator](./glossary#operator).

#### Instance

The term *instance* is used to describe a specific instance of a specific type (usually
[Operator](./glossary#operator) or [Function](./glossary#function)) during runtime. As Apache Flink
is mostly written in Java, this corresponds to the definition of *Instance* or *Object* in Java.
In the context of Apache Flink, the term *parallel instance* is also frequently used to emphasize
that multiple instances of the same [Operator](./glossary#operator) or
[Function](./glossary#function) type are running in parallel.

#### Flink Job

A Flink Job is the runtime representation of a Flink program. A Flink Job can either be submitted
to a long running [Flink Session Cluster](./glossary#flink-session-cluster) or it can be started as a
self-contained [Flink Application Cluster](./glossary#flink-application-cluster).

#### JobGraph

see [Logical Graph](./glossary#logical-graph)

#### Flink JobManager

JobManagers are one of the components running in the

> [Flink JobManger Process](./glossary#flink-jobmanager-process). A JobManager is responsible for

Reviewer: `JobManger` -> `JobManager`

Author: Done.

#### Operator

Node of a [Logical Graph](./glossary#logical-graph). An Operator performs a certain operation,
which is usually executed by a [Function](./glossary#function). Sources and Sinks are special
Operators for data ingestion and data egress.

#### Operator Chain

An Operator Chain consists of two or more consecutive [Operators](./glossary#operator) without any
repartitioning in between. Operators within the same Operation Chain forward records to each other
directly without going through serialization or Flink's network stack.

#### Partition

A partition is an independent subset of the overall data stream or data set. A data stream or
data set is divided into partitions by assigning each [record](./glossary#Record) to one or more
partitions. Partitions of data streams or data sets are consumed by [Tasks](./glossary#task) during
runtime. A transformation which changes the way a data stream or data set is partitioned is often
called repartitioning.

#### Physical Graph

A physical graph is the result of translating a [Logical Graph](./glossary#logical-graph) for
execution in a distributed runtime. The nodes are [Tasks](./glossary#task) and the edges indicate

> input/output-relationships or [partitions](./glossary#partition) of data streams or data sets.

Reviewer: "physical graph" somehow implies this is the runtime graph of task instances on the nodes, but this is still only a data structure of how to deploy things, right?

Author: That's a good question and also related to your other comment below. I would have said that the "physical graph" indeed is the "runtime graph" of task instances. Of course it is just a model of the actual physical graph, but it models the execution during runtime.

#### Record

Records are the constituent elements of a data set or data stream.
[Operators](./glossary#operator) and [Functions](./glossary#Function) receive records as input
and emit records as output.

#### Flink Session Cluster

A long-running [Flink Cluster](./glossary#flink-cluster) which accepts multiple
[Flink Jobs](./glossary#flink-job) for execution. The lifetime of this Flink Cluster is not bound
to the lifetime of any Flink Job. Formerly, a Flink Session Cluster was also known as a Flink Cluster in
*session mode*. Compare to [Flink Application Cluster](./glossary#flink-application-cluster).

#### State Backend

> For stream processing programs, the State Backend determines how state is stored on each Taskmanager

Reviewer: `TaskManager`

Author: Done.

> (Java Heap of Taskmanager or (embedded) RocksDB) as well as where it is written upon a checkpoint

Reviewer: `TaskManager`

(Java Heap of [Flink JobManager Process](./glossary#flink-jobmanager-process) or Filesystem).

#### Sub-Task

> A Sub-Task is an instance of [Task](./glossary#task) responsible for processing a

Reviewer: `of a [Task]...`

Author: Done.

[partition](./glossary#partition) of the data stream. The term "Sub-Task" emphasizes that there are

> multiple parallel Tasks for the same [Operator](./glossary#operator).

Reviewer: `[Operator](./glossary#operator) or [Operator Chain](./glossary#operator-chain)`

Author: Done.

#### Task

Node of a [Physical Graph](./glossary#physical-graph). A task is the basic unit of work, which is
executed by Flink's runtime. Tasks encapsulate exactly one parallel instance of an

> [Operator](./glossary#operator) or [Operator Chain](./glossary#operator-chain).

Reviewer: You got me confused a but with the last sentence thinking that a Task is also an instance. How about this:
`Tasks logically encapsulate exactly one parallel instance...`

#### Flink TaskManager

TaskManagers are the worker processes of a [Flink Cluster](./glossary#flink-cluster).
[Tasks](./glossary#task) are scheduled to TaskManagers for execution. They communicate with each
other to exchange data between subsequent Tasks.

#### Transformation

A Transformation is applied on one or more data streams or data sets and results in one or more
output data streams or data sets. A transformation might change a data stream or data set on a

> per-record basis, but but might also only change its partitioning or perform an aggregation. While

Reviewer: `but but` -> `but`

Author: Done.

> [Operators](./glossary#operator) and [Functions](./glossary#function)) are "physical" parts Flink's

Reviewer: `are the "physical" parts of Flink's`

Author: Done.

> API, Transformations are only an API concept. Specifically, most - but not all -  transformations

Reviewer: there is one space too many after the last "-"

Author: Done.

or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

#### Flink Application Cluster

A Flink Application Cluster is a dedicated [Flink Cluster](./glossary#flink-cluster) that only
executes a single [Flink Job](./glossary#flink-job). The lifetime of the
[Flink Cluster](./glossary#flink-cluster) is bound to the lifetime of the Flink Job. Formerly
Flink Application Clusters were also known as Flink Clusters in *job mode*. Compare to
[Flink Session Cluster](./glossary#flink-session-cluster).

#### Flink Cluster

> The distributed system consisting of (typically) one Flink Master process and one or more Flink

Reviewer 6: Suggestion: 
 * `The distributed system ...` -> `A distributed system ...`
 * `Flink Master process` -> `[Flink JobManager process](#flink-jobmanager-process)`
 * `Flink TaskManager processes` -> `[Flink TaskManager processes](#...)`

Author: 2) and 3) done. 1) I am not sure.Maybe let's ask a native speaker. @sjwiesman what sounds more natural to you?

Reviewer 7: "A distributed system"

Reviewer 7: Also, drop the (typically) ?

Author: Standby JobManager Processes? I would considers these to be part of the Flink Cluster.

Reviewer 7: Gotcha, maybe as a nit "The distributed system consisting of one Flink Master process, zero or more stand by job manager processes, and one or more Flink . . ."

Otherwise, lets get this merged in. We can always open follow up pull requests.

#### Logical Graph

A logical graph is a directed graph describing the high-level logic of a stream processing program.
The nodes are [Operators](./glossary#operator) and the edges indicate input/output-relationships or
data streams or data sets.

#### Managed State

Managed State describes application state which has been registered with the framework. For
Managed State, Apache Flink will take care about persistence and rescaling among other things.

#### Flink JobManager Process

The JobManager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called
*JobManager* for historical reasons, but actually contains three distinct components:
Flink Resource Manager, Flink Dispatcher and one [Flink JobManager](./glossary#flink-jobmanager)
per running [Flink Job](./glossary#flink-job).

#### Operator

Node of a [Logical Graph](./glossary#logical-graph). An Operator performs a certain operation,
which is usually executed by a [Function](./glossary#function). Sources and Sinks are special
Operators for data ingestion and data egress.

#### Operator Chain

An Operator Chain consists of two or more consecutive [Operators](./glossary#operator) without any

> repartitioning in between. Operators within the same Operation Chain forward records to each other

Reviewer 6: Typo: `Operation Chain` -> `Operator Chain`.

Author: Done

partitions. Partitions of data streams or data sets are consumed by [Tasks](./glossary#task) during
runtime. A transformation which changes the way a data stream or data set is partitioned is often
called repartitioning.

#### Physical Graph

A physical graph is the result of translating a [Logical Graph](./glossary#logical-graph) for
execution in a distributed runtime. The nodes are [Tasks](./glossary#task) and the edges indicate
input/output-relationships or [partitions](./glossary#partition) of data streams or data sets.

#### Record

Records are the constituent elements of a data set or data stream.
[Operators](./glossary#operator) and [Functions](./glossary#Function) receive records as input
and emit records as output.

#### Flink Session Cluster

A long-running [Flink Cluster](./glossary#flink-cluster) which accepts multiple
[Flink Jobs](./glossary#flink-job) for execution. The lifetime of this Flink Cluster is not bound
to the lifetime of any Flink Job. Formerly, a Flink Session Cluster was also known as a Flink Cluster in
*session mode*. Compare to [Flink Application Cluster](./glossary#flink-application-cluster).

#### State Backend

> For stream processing programs, the State Backend determines how state is stored on each TaskManager

Reviewer 6: `the State Backend determines how state is stored` -> `a State Backend determines how Flink Job's state is stored`?

In general, it's a little bit not clear from here, that a state backend can be configured on Flink job level.

Author: Should be clearer now.

(Java Heap of TaskManager or (embedded) RocksDB) as well as where it is written upon a checkpoint
(Java Heap of [Flink JobManager Process](./glossary#flink-jobmanager-process) or Filesystem).

#### Sub-Task

A Sub-Task is an instance of a [Task](./glossary#task) responsible for processing a
[partition](./glossary#partition) of the data stream. The term "Sub-Task" emphasizes that there are
multiple parallel Tasks for the same [Operator](./glossary#operator) or
[Operator Chain](./glossary#operator-chain).

#### Task

Node of a [Physical Graph](./glossary#physical-graph). A task is the basic unit of work, which is
executed by Flink's runtime. Tasks logically encapsulate exactly one parallel instance of an
[Operator](./glossary#operator) or [Operator Chain](./glossary#operator-chain).

> #### Flink TaskManager

Reviewer 6: Can be `Flink TaskManager Process` to be similar with `Flink JobManager Process`.

Author: We only introduced the term "Flink JobManager Process" to differentiate between "Flink JobManager" (a component) of the "Flink JobManager Process". I think for TaskManager we don't need this.

Author: As discussed with Till and Stephan we went for "Flink Master" after all.

TaskManagers are the worker processes of a [Flink Cluster](./glossary#flink-cluster).
[Tasks](./glossary#task) are scheduled to TaskManagers for execution. They communicate with each
other to exchange data between subsequent Tasks.

#### Transformation

A Transformation is applied on one or more data streams or data sets and results in one or more
output data streams or data sets. A transformation might change a data stream or data set on a
per-record basis, but might also only change its partitioning or perform an aggregation. While
[Operators](./glossary#operator) and [Functions](./glossary#function)) are the "physical" parts

> Flink's API, Transformations are only an API concept. Specifically, most - but not all -

Reviewer 6: `Flink's API` -> `of Flink's API`?

Author: Done

> transformations are implemented by a certain [Operator](./glossary#operator).

Reviewer 6: `by a certain [Operator]` -> `by certain [Operators]`?

Author: Done
